File: Data Structures & Algorithms/combination-target-sum/submission-4.py
class Solution:
    def __init__(self):
        self.out = []
        self.cur = []
    def _dfs(self, nums: List[int], idx: int, target: int):
        if target == 0:
            self.out.append(self.cur[:])
            return
        
        if target < 0 or idx == len(nums):
            return
        
        self.cur.append(nums[idx])
        #Take all cases where current value is included
        self._dfs(nums, idx, target-nums[idx])
        self.cur.pop()
        # Take case not including current value
        self._dfs(nums, idx+1, target)

    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        self._dfs(nums, 0, target)
        return self.out        

File: Data Structures & Algorithms/climbing-stairs/submission-4.cpp
class Solution {
public:
    int dfs(int n, vector<int>&cache)
    {
        if(cache[n] != -1)
            return cache[n];

        if(n == 0)
            return 1;
        
        if(n < 0)
            return 0;

        cache[n] = dfs(n-1, cache) + dfs(n-2, cache);
        return cache[n];
    }
    int climbStairs(int n) {
        vector<int>cache(n+1, -1);
        return dfs(n, cache);        
    }
};
